### ISSUE TITLE
"pg: no data source matches prefix" error

### ISSUE THREAD

**Issue by florence-crl** [cockroachdb/cockroach#43963]
**Describe the problem**
A user is getting this error in v19.1.3:
```
with a as (select 1 as testval) select a.testval from a;
pq: no data source matches prefix: a
```
This statement does not give the error in  v19.2.2, however it still gives the error in 19.1.6. 

**To Reproduce**

Using v19.1.3 and v19.1.6,
```
root@localhost:26267/defaultdb> with a as (select 1 as testval) select a.testval from a;
pq: no data source matches prefix: a
```
**Expected behavior**
I expected v19.1.6 to get the results below, that I got from using 19.2.2
```
root@localhost:26257/defaultdb> with a as (select 1 as testval) select a.testval from a;
  testval
+---------+
        1
(1 row)

Time: 992µs
```
**Additional data / screenshots**

**Environment:**
 - CockroachDB version 19.1.6
 - Server OS: 
 - Client app `cockroach sql`

**Additional context**
This issue is preventing user from using the Looker BI tool, as it is generating Postgres-compatible SQL that does not work with CockroachDB. They do not have control over the generated SQL. This is impacting their ability to deliver BI solutions to their clients.

**Comment by RoachietheSupportRoach** [comment]
Zendesk ticket #[4413](https://cockroachdb.zendesk.com/agent/tickets/4413) has been linked to this issue.

**Comment by RaduBerinde** [comment]
@rytaft, could you take a look? Might just be a missing backport

**Comment by rytaft** [comment]
https://github.com/cockroachdb/cockroach/pull/38670 is the PR that would fix this issue, but it may be too risky for a backport, since it's a relatively big change (adding the `With` operator in the optimizer) and there were a number of issues with the `With` operator that were fixed by later PRs.

**Comment by rytaft** [comment]
#45782 is merged, so it should be available in the next patch release of 19.1. Closing this.
